File: Data Structures & Algorithms/n-queens/submission-4.java
class Solution {
    int n;
    List<List<String>> res;
    List<String> board;
    Set<Integer> col;
    Set<Integer> diagRight;
    Set<Integer> diagLeft;
    
    private void dfs(int i) {
        if (i == n) {
            res.add(new ArrayList<>(board));
            return;
        }

        StringBuilder sb = new StringBuilder();
        for (int k = 0; k < n; ++k) {
            sb.append('.');
        }

        for (int j = 0; j < n; ++j) {
            if (!col.contains(j) &&
                !diagRight.contains(i - j) &&
                !diagLeft.contains(i + j)) {
                    col.add(j);
                    diagRight.add(i - j);
                    diagLeft.add(i + j);
                    sb.setCharAt(j, 'Q');
                    
                    board.add(sb.toString());
                    dfs(i + 1);
                    board.remove(board.size() - 1);

                    col.remove(j);
                    diagRight.remove(i - j);
                    diagLeft.remove(i + j);
                    sb.setCharAt(j, '.');
            }
        }
    }

    public List<List<String>> solveNQueens(int n) {
        this.n = n;
        res = new ArrayList<>();
        board = new ArrayList<>();
        diagRight = new HashSet<>();
        col = new HashSet<>();
        diagLeft = new HashSet<>();

        dfs(0);

        return res;
    }
}

File: Data Structures & Algorithms/count-subsequences/submission-4.py
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        m = len(s)
        n = len(t)

        dp = {}

        def dfs(i, j):
            if (i, j) in dp:
                return dp[(i, j)]
            elif j == n:
                return 1
            elif i == m:
                return 0

            count = 0

            for k in range(i, m):
                if s[k] == t[j]:
                    count += dfs(k + 1, j + 1)

            dp[(i, j)] = count
            return dp[(i, j)]

        return dfs(0, 0)